content/ui-ux.md

```diff
@@ -0,0 +1,514 @@
+
+#### Scroll events
+
+Another type of event that fires very frequently is a scroll event. Such event trigger as a user scrolls up and down a page and are commonly used for "waypoint" behaviour. For example in the Localmarket example app, we use the [jquery-waypoints](https://atmospherejs.com/meteor/jquery-waypoints) library to trigger a class being attached to the nav element when you scroll past a certain point:
+
+```js
+Template.nav.onRendered(function() {
+  var $nav = this.$('nav');
+  $nav.siblings('.content-scrollable:not(.static-nav)').children().first().waypoint(function(direction) {
+    $nav.toggleClass('scrolled', direction === 'down');
+  }, {
+    context: '.content-scrollable',
+    offset: -200
+  });
+});
+```
+
+An important thing to be careful of is that if you are not using a library such as `jquery-waypooints` to do this, then you need to be careful about not causing too much work to happen in each `scroll` event. If you do so, then users will see poor scrolling performance and your app will not feel performant. You can use a throttle or [`requestAnimationFrame`](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame) to make things smooth if you are careful.
+
+
+## User Experience Patterns
+
+There are some common user experience (UX---as in how your app behaves) patterns that are typical to most Meteor apps that are worth exploring here. These patterns relate heavily to the way the data they are allow the user is interacting with is subscribed to and published, so there are similar sections in the {data loading article} which deal with the data side of them.
+
+### Subscription readiness
+
+When you subscribe to data in Meteor, it does not become instantly available on the client. Typically the user will need to wait for at least a few hundred milliseconds (or as long as a few seconds) for the data to arrive. This is especially noticable when the app first starts up or you move between screens.
+
+There are a few UX techniques for dealing with this waiting period. The simplest is simply to switch out the page you are rendering with a generic "loading" page while you wait for all the data (typically a page may open several subscriptions) to load. As an example, in the Todos example app, we wait until all the public lists and the user's private lists have loaded before we try to render the actual page:
+
+```blaze
+{{#if Template.subscriptionsReady}}
+  {{> Template.dynamic template=main}}
+{{else}}
+  {{> appLoading}}
+{{/if}}
+```
+
+We do this with Blaze's `Template.subscriptionsReady` which is perfect for this purposes, as it waits for all the subscriptions that the current component has asked for to become ready.
```

Comment:
I think we do cross link pretty heavily. It's subtle -- the data loading chapter talks about what the pubs should look like and where you should subscribe, this chapter talks about how you should build your UI based on the complexity of information you get back from it.